crates/transpiler/src/passes/pad_dynamical_decoupling.rs

```diff
@@ -0,0 +1,515 @@
+                if let Some((inv_gate, _inv_params)) = gate.inverse(&gate_params) {
+                    let inv_packed_op = PackedOperation::from_standard_gate(inv_gate);
+                    inv_packed_op
+                        .matrix(&[])
+                        .expect("No matrix representation for inverse gate.")
+                } else {
+                    panic!("No inverse for this standard gate!");
+                }
+            }
+            OperationRef::Gate(py_gate) => {
+                // Extract the inverse from the Python gate
+                let inv_gate: PyObject = py_gate.gate.call_method0(py, "inverse")?;
+                let inv_gate_rust: PackedOperation = inv_gate.extract::<OperationFromPython>(py).unwrap().operation.clone();
+                let inv_matrix: Array2<Complex64> = inv_gate_rust
+                    .matrix(&[])
+                    .expect("No matrix representation for inverse gate.");
+                inv_matrix
+            }
+            _ => unreachable!("Only Python Gate types (standard/non-standard) should be passed into this function."),
+        };
+
+        let [theta, phi, lam, phase] = angles_from_unitary(u_inv.view(), EulerBasis::U3);
+
+        if next_node.is_instance_of::<DAGOpNode>() {
+            let mut next_node_ref: PyRefMut<DAGOpNode> = next_node.extract()?;
+            let next_node_inst: &mut CircuitInstruction = &mut next_node_ref.instruction;
+            let next_node_op: &PackedOperation = &next_node_inst.operation;
+
+            // Check if the next node corresponds to a U/U3 gate
+            if let Some(gate) = next_node_op.try_standard_gate() {
+                match gate {
+                    StandardGate::U | StandardGate::U3 => {
+                        // Absorb the inverse into the successor (from right in circuit)
+                        let [theta_r_param, phi_r_param, lam_r_param]: [Param; 3] =
+                            next_node_inst.params.clone().into_inner().ok().unwrap();
+                        let mut theta_r: f64 = 0.;
+                        let mut phi_r: f64 = 0.;
+                        let mut lam_r: f64 = 0.;
+                        if let Param::Float(theta) = theta_r_param {
+                            theta_r = theta;
+                        }
+                        if let Param::Float(phi) = phi_r_param {
+                            phi_r = phi;
+                        }
+                        if let Param::Float(lam) = lam_r_param {
+                            lam_r = lam;
+                        }
+
+                        let [theta_new, phi_new, lam_new] =
+                            compose_u3_rust(theta_r, phi_r, lam_r, theta, phi, lam);
+
+                        next_node_inst.params = smallvec![
+                            Param::Float(theta_new),
+                            Param::Float(phi_new),
+                            Param::Float(lam_new),
+                        ];
```

Comment:
This likely won't work, since a `CircuitInstruction` on a `DAGNode` would be ephemeral (I think). You might need to replace the node in the DAG instead for this to work.